File: Path Sum B-Tree/Java/solution1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

     public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public boolean hasPathSum(TreeNode root, int sum) { 
            return pathSumHelper(root,0,sum); 
    }
    
   ///**
   // pathSumHelper recursively finds a sum path
   // root: node of tree
   // curSum current sum of the path
   // target the sum to reach
   ///
    public boolean pathSumHelper(TreeNode root, int curSum, int target){
        if(root == null) return false;
        // if the sum plus the node val is the target and the node is a leaf return true
        if(curSum+root.val == target && root.left == null && root.right == null) return true;

        // check both branches of the node for a path. If either is true return
        boolean path = pathSumHelper(root.left, curSum+root.val, target) || pathSumHelper(root.right, curSum+root.val, target);
        
        return path;
    }
}

File: Path Sum B-Tree/Python/solution1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        return self.pathSumHelper(root, targetSum, 0)
    
    def pathSumHelper(self, node, target, runSum):
        if node is None:
            return False
        # if the node is a leaf return if the sum is reached
        if node.left is None and node.right is None:
            runSum += node.val
            if runSum == target:
                return True
        # the node is not a leaf so add to the running sum
        runSum += node.val
        ## check the left and right branches of the tree and return true as soon as one is found
        ok = self.pathSumHelper(node.left, target, runSum)
        if ok:
            return True
        ok = self.pathSumHelper(node.right, target, runSum)
        if ok:
            return True
        
        return False